### PR TITLE
[swift5] Add property `x-null-encodable` extension for full control over encoding value/nil or nothing

### PR DESCRIPTION
**Pull request by jarrodparkes**
Attempts to solve #11033

This PR adds a new property vendor extension `x-null-encodable` for use with the `swift5` generator. If set to true, then the generated properties will use the `NullEncodable<Wrapped>` type:

```swift
public enum NullEncodable<Wrapped> {
    case encodeNothing
    case encodeNull
    case encodeValue(Wrapped)
}
```

This type gives the client full control of whether to encode a value/null -OR- encode nothing.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under ./bin/ (or Windows batch scripts under.\bin\windows) to update Petstore samples related to your fix. This is important, as CI jobs will verify all generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run ./bin/{LANG}-petstore.sh, ./bin/openapi3/{LANG}-petstore.sh if updating the code or mustache templates for a language ({LANG}) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): master, 4.3.x, 5.0.x. Default: master.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Core Team Members

@jgavris 
@ehyche
@Edubits 
@jaz-ah 
@4brunu

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/swift5/modelObject.mustache [hunk 1]
```diff
@@ -11,12 +11,12 @@
 {{#isEnum}}
     {{#description}}/** {{{.}}} */
     {{/description}}{{#deprecated}}@available(*, deprecated, message: "This property is deprecated.")
-    {{/deprecated}}{{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}public{{/nonPublicApi}} {{#readonlyProperties}}private(set) {{/readonlyProperties}}var {{{name}}}: {{{datatypeWithEnum}}}{{#required}}{{#isNullable}}?{{/isNullable}}{{/required}}{{^required}}?{{/required}}{{#defaultValue}} = {{{.}}}{{/defaultValue}}
+    {{/deprecated}}{{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}public{{/nonPublicApi}} {{#readonlyProperties}}private(set) {{/readonlyProperties}}var {{{name}}}: {{#vendorExtensions.x-null-encodable}}NullEncodable<{{{datatypeWithEnum}}}>{{/vendorExtensions.x-null-encodable}}{{^vendorExtensions.x-null-encodable}}{{{datatypeWithEnum}}}{{#required}}{{#isNullable}}?{{/isNullable}}{{/required}}{{^required}}?{{/required}}{{#defaultValue}} = {{{.}}}{{/defaultValue}}{{/vendorExtensions.x-null-encodable}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/swift5/modelObject.mustache:14]: @jarrodparkes I think here you closed the {{/vendorExtensions.x-null-encodable}} to late, and it should be after datatypeWithEnum, maybe something like this?

`{{/deprecated}}{{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}public{{/nonPublicApi}} {{#readonlyProperties}}private(set) {{/readonlyProperties}}var {{{name}}}: {{#vendorExtensions.x-null-encodable}}NullEncodable<{{{datatypeWithEnum}}}>{{/vendorExtensions.x-null-encodable}}{{^vendorExtensions.x-null-encodable}}{{{datatypeWithEnum}}}{{/vendorExtensions.x-null-encodable}}{{#required}}{{#isNullable}}?{{/isNullable}}{{/required}}{{^required}}?{{/required}}{{#defaultValue}} = {{{.}}}{{/defaultValue}}`

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/swift5/modelObject.mustache:14]: Or maybe not and this is this way by design?
But aren't we losing the optional and default value information?

File: samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift [hunk 2]
```diff
@@ -14,11 +14,11 @@ import AnyCodable
 public struct Name: Codable, Hashable {
 
     public var name: Int
-    public var snakeCase: Int?
+    public var snakeCase: NullEncodable<Int>
     public var property: String?
     public var _123number: Int?
 
-    public init(name: Int, snakeCase: Int? = nil, property: String? = nil, _123number: Int? = nil) {
+    public init(name: Int, snakeCase: NullEncodable<Int>, property: String? = nil, _123number: Int? = nil) {
```
Comment (4brunu) [inline samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: @jarrodparkes this is what I was talking about above, shouldn't we keep the optional and default value information?

Comment (4brunu) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: Maybe something like `snakeCase: NullEncodable<Int?> = .encodeValue(nil)`

Comment (jarrodparkes) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: @4brunu 

yeh, I wasn't sure if I had enough information from the mustache file to determine whether the default should be...

```swift
snakeCase: NullEncodable<Int> = .encodeNull
// or
snakeCase: NullEncodable<Int> = .encodeNothing
// or
snakeCase: NullEncodable<Int> = .encodeSome(...)
```

How would you go about this?

**Comment by 4brunu** [comment]
@jarrodparkes one question, this only applies to properties that can be nil or applies to all properties?

File: modules/openapi-generator/src/main/resources/swift5/modelObject.mustache (hunk 1 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/modelObject.mustache:14]: @4brunu We aren't losing the optional information insofar as the new type that is being used (`NullEncodable`) can encode null/nothing. We are losing the default value information. See my question below.

File: samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift (hunk 2 above)
Comment (jarrodparkes) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: > NullEncodable<Int?>

Above, the optional type in <brackets> is redundant. I was removing the `?` by design, but let me know what you think.

**Comment by jarrodparkes** [comment]
@4brunu

> this only applies to properties that can be nil or applies to all properties?

This behavior applies to properties earmarked as `x-null-encodable` in the spec. Properties with this extension are encoded into the new `NullEncodable` type which gives clients the flexibility to encode a value, encode null, or encode nothing for the property. This is different than our current behaviors...

1. A property that is `nullable` and `required` will always be encoded regardless of its value
2. A property that is `nullable` will only be encoded when the value is non-nil

**Comment by 4brunu** [comment]
So this will basically override/ignore the `nullable` and `required`, right?

**Comment by jarrodparkes** [comment]
@4brunu yeh, more or less. the spec example doesn't make that clear because `x-null-encodable` is used alongside `nullable` and `required`. but, to your point, `x-null-encodable` essentially overrides them

File: samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift (hunk 2 above)
Comment (4brunu) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: I'm not sure if that information is available or not, but looking at https://github.com/OpenAPITools/openapi-generator/issues/7521 I would say that maybe it's not available.
So maybe it will be hard to support default values

Comment (jarrodparkes) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: @4brunu 

As part of the `Swift5ClientCodegen` could we do something in `postProcessModels` to detect the extension and override/set the default value?

Something like this...

| Previous Default Value | New Default Value |
|-|-|
|nil|.encodeNull|
|5|.encodeValue(5)|
|{literal}|.encodeValue({literal})|

Comment (4brunu) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: That's a good idea! I was about to give up on the default value.

Comment (jarrodparkes) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: @4brunu I'm trying to add something like _this_ to `postProcessModels`...

```java
                if (prop.name.equals("x-null-encodable")) {
                    if (null == prop.defaultValue) {
                        prop.setDefaultValue(".encodeNull");
                    } else {
                        prop.setDefaultValue(".encodeValue(" + prop.defaultValue + ")");
                    }
                }
```

But it seems like this doesn't take effect or might be ignored. Maybe `toDefaultValue` overrides it?

Comment (4brunu) [reply samples/client/petstore/swift5/alamofireLibrary/PetstoreClient/Classes/OpenAPIs/Models/Name.swift:21]: Maybe, but I don't have experience with that part of the generator

**Comment by jarrodparkes** [comment]
@4brunu I was able to restore the default values 👍 let me know what you think

File: modules/openapi-generator/src/main/resources/swift5/model_doc.mustache [hunk 3]
```diff
@@ -3,7 +3,7 @@
 ## Properties
 Name | Type | Description | Notes
 ------------ | ------------- | ------------- | -------------
-{{#vars}}**{{{name}}}** | {{#isPrimitiveType}}**{{{dataType}}}**{{/isPrimitiveType}}{{^isPrimitiveType}}{{^isContainer}}[**{{dataType}}**]({{complexType}}.md){{/isContainer}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{/isPrimitiveType}} | {{description}} | {{^required}}[optional] {{/required}}{{#isReadOnly}}[readonly] {{/isReadOnly}}{{#defaultValue}}[default to {{{.}}}]{{/defaultValue}}
+{{#vars}}**{{{name}}}** | {{#isPrimitiveType}}**{{{dataType}}}**{{/isPrimitiveType}}{{^isPrimitiveType}}{{^isContainer}}[**{{dataType}}**]({{complexType}}.md){{/isContainer}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{/isPrimitiveType}} | {{description}} | {{^required}}[optional] {{/required}}{{#isReadOnly}}[readonly] {{/isReadOnly}}{{#defaultValue}}[default to {{#vendorExtensions.x-null-encodable}}.encodeValue({{{.}}}){{/vendorExtensions.x-null-encodable}}{{^vendorExtensions.x-null-encodable}}{{{.}}}{{/vendorExtensions.x-null-encodable}}]{{/defaultValue}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @jarrodparkes here should we check if the default value is null and put .encondeNull?

File: modules/openapi-generator/src/test/resources/2_0/swift/petstore-with-fake-endpoints-models-for-testing.yaml [hunk 4]
```diff
@@ -1212,6 +1214,7 @@ definitions:
         readOnly: true
         type: integer
         format: int32
+        x-null-encodable: true
```
Comment (4brunu) [inline modules/openapi-generator/src/test/resources/2_0/swift/petstore-with-fake-endpoints-models-for-testing.yaml:1217]: @jarrodparkes it would be nice to have a property that is nullable with the default value of null with `x-null-encodable: true`, to make the test case broader

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/Swift5ClientCodegen.java [hunk 5]
```diff
@@ -1085,6 +1085,16 @@ public String toEnumName(CodegenProperty property) {
                     prop.vendorExtensions.put("x-codegen-escaped-property-name", true);
                     modelHasPropertyWithEscapedName = true;
                 }
+
+                if (prop.vendorExtensions.containsKey("x-null-encodable")) {
+                    if (prop.vendorExtensions.get("x-null-encodable").toString().equals("true")) {
+                        if (prop.defaultValue == null || prop.defaultValue.equals("null")) {
+                            prop.vendorExtensions.put("x-codegen-alt-default-value", ".encodeNull");
```
Comment (4brunu) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/Swift5ClientCodegen.java:1092]: @jarrodparkes what about renaming this variable to `x-null-encodable-default-value`?

**Comment by 4brunu** [comment]
@jarrodparkes overall this looks good to me, I just left some small comments on some improvements/questions

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/Swift5ClientCodegen.java (hunk 5 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/Swift5ClientCodegen.java]: fixed

File: modules/openapi-generator/src/test/resources/2_0/swift/petstore-with-fake-endpoints-models-for-testing.yaml (hunk 4 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/test/resources/2_0/swift/petstore-with-fake-endpoints-models-for-testing.yaml:1217]: fixed

File: modules/openapi-generator/src/main/resources/swift5/model_doc.mustache (hunk 3 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @4brunu I'm not sure what is different about this template. From this template, it is like I am unable to access {{{x-null-encodable-default-value}}}. Maybe this template is resolved prior to the `postProcessModels` function call?

I'm not sure there is a mustache check I can make which checks IF defaultValue == null.

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: Hum, I see, since this is in the docs, I don't think it's too bad.

**Comment by 4brunu** [comment]
CI is failing, but it looks like it's not related to this PR.
Could you please close and reopen this PR please, to restart the CI?

**Comment by jarrodparkes** [comment]
@4brunu sure thing

File: modules/openapi-generator/src/main/resources/swift5/model_doc.mustache (hunk 3 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @4brunu Is there some sort of `preProcessModels` method? I'm just wondering if I am on to something about when this template is resolved in relation to the `postProcessModels` function

Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: weird... looking into `DefaultGenerator` it seems like `postProcessModels` would have already been called...

https://github.com/OpenAPITools/openapi-generator/blob/611a3effa60d32b59c9402766a87d144ecf1eec4/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L507#L541

```java
        allProcessedModels = config.postProcessAllModels(allProcessedModels);


        // generate files based on processed models
        for (String modelName : allProcessedModels.keySet()) {
            Map<String, Object> models = (Map<String, Object>) allProcessedModels.get(modelName);
            models.put("modelPackage", config.modelPackage());
            try {
                //don't generate models that have an import mapping
                if (config.importMapping().containsKey(modelName)) {
                    continue;
                }


                // TODO revise below as we've already performed unaliasing so that the isAlias check may be removed
                List<Object> modelList = (List<Object>) models.get("models");
                if (modelList != null && !modelList.isEmpty()) {
                    Map<String, Object> modelTemplate = (Map<String, Object>) modelList.get(0);
                    if (modelTemplate != null && modelTemplate.containsKey("model")) {
                        CodegenModel m = (CodegenModel) modelTemplate.get("model");
                        if (m.isAlias && !(config instanceof PythonClientCodegen))  {
                            // alias to number, string, enum, etc, which should not be generated as model
                            // for PythonClientCodegen, all aliases are generated as models
                            continue;  // Don't create user-defined classes for aliases
                        }
                    }
                    allModels.add(modelTemplate);
                }


                // to generate model files
                generateModel(files, models, modelName);


                // to generate model test files
                generateModelTests(files, models, modelName);


                // to generate model documentation files
                generateModelDocumentation(files, models, modelName);
```

Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: maybe it is an issue with the mustache file accessing the value 🤔 ? where that line is located is wrapped by `{{#vars}}{{/vars}}` instead of `{{#allVars}}{{/allVars}}`... would that have anything to do with it?

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: In this area I can't help much because because I never touch that area before :/

**Comment by 4brunu** [comment]
That's weird the CI failed again with a different error this time, that is still not related to this PR (I think)

File: modules/openapi-generator/src/main/resources/swift5/model_doc.mustache (hunk 3 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @4brunu do you know anyone who might be able to point us in the right direction?

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: Maybe @wing328 can help with this?

Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @wing328 TL;DR

In `postProcessModels` we inject a custom property vendorExtension called `x-null-encodable-default-value`. However, that value doesn't appear to be present/usable in the `model_doc.mustache` template.

**Comment by jarrodparkes** [comment]
@4brunu yeh, seems like that might be an issue across a few different PRs. 

@wing328 still curious, if you might have some insights on the following...

TL;DR - In `postProcessModels` we inject a custom property vendorExtension called `x-null-encodable-default-value`. However, `x-null-encodable-default-value` doesn't appear to be present/usable in the `model_doc.mustache` template. Looking at `DefaultGenerator.java` it seems that the model documentation is resolved AFTER `postProcessModels`, so perhaps we aren't using something correctly in the mustache template? Any help would be greatly appreciated.

File: modules/openapi-generator/src/main/resources/swift5/model_doc.mustache (hunk 3 above)
Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: @4brunu bumping this incase you have any different thoughts. hope you are having a Merry Christmas season 🎄

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: Hi @jarrodparkes, sorry but I don't have any ideas.
But since this only affects documentation, maybe it's okay to leave it like that

Comment (jarrodparkes) [reply modules/openapi-generator/src/main/resources/swift5/model_doc.mustache:6]: that works for me

**Comment by jarrodparkes** [comment]
@4brunu let me know if there anything else you need to see here. AFAIK the failed check isn't related to this PR

**Comment by wing328** [comment]
Restarted the bitrise job. Let's see how that goes.

**Comment by wing328** [comment]
bitrise job passed via https://app.bitrise.io/build/6c82ea06-8680-4bd4-8079-b94726e9ea98

**Comment by 4brunu** [comment]
Looks good to me 👍

**Comment by jarrodparkes** [comment]
anything holding this up @4brunu @wing328 ?
